lib/spack/spack/package.py

```diff
@@ -1505,24 +1519,26 @@ def do_install(self, **kwargs):
             # module from the upstream Spack instance.
             return
 
-        partial = self.check_for_unfinished_installation(keep_prefix, restage)
-
-        # Ensure package is not already installed
-        layout = spack.store.layout
-        with spack.store.db.prefix_read_lock(self.spec):
-            if partial:
-                tty.msg(
-                    "Continuing from partial install of %s" % self.name)
-            elif layout.check_installed(self.spec):
-                msg = '{0.name} is already installed in {0.prefix}'
-                tty.msg(msg.format(self))
-                rec = spack.store.db.get_record(self.spec)
-                # In case the stage directory has already been created,
-                # this ensures it's removed after we checked that the spec
-                # is installed
-                if keep_stage is False:
-                    self.stage.destroy()
-                return self._update_explicit_entry_in_db(rec, explicit)
+        if self.has_code:
+            # Ensure package code is not already installed
+            partial = self.check_for_unfinished_installation(keep_prefix,
+                                                             restage)
```

Comment:
This is probably also be too strict, and possibly wrong.  If the package was *previously* some other type of package (i.e. one that `has_code`) then there may well be an existing installation that we didn't stick in the DB, and this will miss it.  I think we should still check for unfinished installations even if `not self.has_code`.

Reply:
Please clarify.  Do you expect someone to change the package type in the "middle" of  an installation (i.e, one that has not completed successfully)?

Reply:
Yep.  Especially while creating a package for the first time.  Someone might well create a package, then decide that it needs to be a bundle package instead and resume work on it.

Reply:
Ah.  Thanks for the use case.